qiskit/qasm/pygments/lexer.py

```diff
@@ -13,96 +13,99 @@
-from pygments.style import Style
-
-
-class QasmTerminalStyle(Style):
-    """A style for OpenQasm in a Terminal env (e.g. Jupyter print)
-    """
-    styles = {
-        String:              'ansibrightred',
-        Number:              'ansibrightcyan',
-        Keyword.Reserved:    'ansibrightgreen',
-        Keyword.Declaration: 'ansibrightgreen',
-        Keyword.Type:        'ansibrightmagenta',
-        Name.Builtin:        'ansibrightblue',
-        Name.Function:       'ansibrightyellow'}
-
-
-class QasmHTMLStyle(Style):
-    """A style for OpenQasm in a HTML env (e.g. Jupyter widget)
-    """
-    styles = {
-        String:              'ansired',
-        Number:              'ansicyan',
-        Keyword.Reserved:    'ansigreen',
-        Keyword.Declaration: 'ansigreen',
-        Keyword.Type:        'ansimagenta',
-        Name.Builtin:        'ansiblue',
-        Name.Function:       'ansiyellow'}
-
-
-class OpenQASMLexer(RegexLexer):
-    """A pygments lexer for OpenQasm
-    """
-    name = 'OpenQASM'
-    aliases = ['qasm']
-    filenames = ['*.qasm']
-
-    gates = ['id', 'cx', 'x', 'y', 'z', 's', 'sdg', 'h',
-             't', 'tdg', 'ccx', 'rx', 'ry', 'rz',
-             'cz', 'cy', 'ch', 'swap', 'cswap', 'crx',
-             'cry', 'crz', 'cu1', 'cu3', 'rxx', 'rzz',
-             'rccx', 'rcccx', 'u1', 'u2', 'u3']
-
-    tokens = {
-        'root': [
-            (r'\n', Text),
-            (r'[^\S\n]+', Text),
-            (r'//\n', Comment),
-            (r'//.*?$', Comment.Single),
-
-            # Keywords
-            (r'(OPENQASM|include)\b', Keyword.Reserved, 'keywords'),
-            (r'(qreg|creg)\b', Keyword.Declaration),
-
-            # Treat 'if' special
-            (r'(if)\b', Keyword.Reserved, 'if_keywords'),
-
-            # Constants
-            (r'(pi)\b', Name.Constant),
-
-            # Special
-            (r'(barrier|measure|reset)\b', Name.Builtin, 'params'),
-
-            # Gates (Types)
-            ('(' + '|'.join(gates) + r')\b', Keyword.Type, 'params'),
-            (r'[unitary\d+]', Keyword.Type),
-
-            # Functions
-            (r'(gate)\b', Name.Function, 'gate'),
-
-            # Generic text
-            (r"[a-zA-Z_][a-zA-Z0-9_]*", Text, 'index')],
-
-        'keywords': [(r'\s*("([^"]|"")*")', String, '#push'),
-                     (r"\d+", Number, '#push'),
-                     (r'.*\(', Text, 'params')],
-
-        'if_keywords': [(r'[a-zA-Z0-9_]*', String, '#pop'),
-                        (r"\d+", Number, '#push'),
-                        (r'.*\(', Text, 'params')],
-
-        'params': [(r"[a-zA-Z_][a-zA-Z0-9_]*", Text, '#push'),
-                   (r'\d+', Number, '#push'),
-                   (r'(\d+\.\d*|\d*\.\d+)([eEf][+-]?[0-9]+)?', Number, '#push'),
-                   (r'\)', Text)],
-
-        'gate': [(r'[unitary\d+]', Keyword.Type, '#push'),
-                 (r'p\d+', Text, '#push')],
-
-        'index': [(r"\d+", Number, '#pop')]
-        }
+try:
```

Comment:
can you avoid conditioning and indenting large code blocks in this file? 
I think the `qiskit.qasm.pygments.lexer` module can exist non-conditionally, but _imported_ conditionally.

Reply:
The reason I had to do this was because the classes defined in this module are subclasses of pygments classes. Python import behavior is hard to workaround with conditionals (for example importing a package will causes all modules to be imported for `dir()`). We need the module to be importable without pygments installed but because of the subclasses this is not feasible. This seemed like best way to do this. The other option here is to remove  any access to `qiskit.qasm.pygments` package from `qiskit.qasm` (or any other outside path) so that `qiskit.qasm.pygments` is only accessible directly via import and we move the import error up to `__init__`. The only issue with this is if someone does a `qiskit.qasm.pygments` import directly it fails. But that's probably not the end of the world. I'll see what I can come up with.

As an aside ideally I think this should live in pygments. I'll try to find the time to dust off my really old PR there (https://bitbucket.org/birkenfeld/pygments-main/pull-requests/769 ) and port over the things from here.

Reply:
Ok, this should be done in: https://github.com/Qiskit/qiskit-terra/pull/4049/commits/40dbf2ff1ffe3334f76d602eca31028d92019440